common/tor.c

```diff
@@ -0,0 +1,232 @@
+#include <arpa/inet.h>
+#include <assert.h>
+#include <ccan/err/err.h>
+#include <ccan/io/io.h>
+#include <ccan/str/hex/hex.h>
+#include <ccan/tal/str/str.h>
+#include <common/tor.h>
+#include <common/type_to_string.h>
+#include <common/utils.h>
+#include <common/wireaddr.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <wire/wire.h>
+
+
+bool parse_tor_wireaddr(const char *arg,u8 *ip_ld,u16 *port_ld)
+{
+	u16 port;
+	char *ip;
+
+	bool res;
+	tal_t *tmpctx = tal_tmpctx(NULL);
+	res = false;
+	ip = tal_strdup(tmpctx, "127.0.0.1");
+	port = 9050;
+	if (!separate_address_and_port(tmpctx, arg,&ip, &port)) {
+		tal_free(tmpctx);
+		return false;
+	}
+	else
+	{
+		assert(strlen(ip) < 16);
+		memcpy(ip_ld,ip,strlen(ip)+1);
+		(*port_ld) = port;
+		res= true;
+	}
+	tal_free(tmpctx);
+	return res;
+}
+
+
+#define MAX_TOR_COOKIE_LEN 32
+#define MAX_TOR_SERVICE_READBUFFER_LEN 255
+#define MAX_TOR_ONION_V2_ADDR_LEN 16
+#define MAX_TOR_ONION_V3_ADDR_LEN 56
+
+struct tor_service_reaching {
+	struct lightningd *ld;
+	u8 buffer[MAX_TOR_SERVICE_READBUFFER_LEN];
+	char *cookie[MAX_TOR_COOKIE_LEN];
+	u8 *p;
+	bool noauth;
+	size_t hlen;
+};
+
+
+static struct io_plan *io_tor_connect_create_onion_finished(struct io_conn *conn, struct tor_service_reaching *reach)
+{
+    u8 tor_dec_bytes[TOR_V3_ADDRLEN];
+
+	if(reach->hlen ==  MAX_TOR_ONION_V2_ADDR_LEN ) {
+	size_t n = tal_count(reach->ld->wireaddrs);
+	tal_resize(&reach->ld->wireaddrs, n+1);
+	reach->ld->wireaddrs[n].type = ADDR_TYPE_TOR_V2;
+	reach->ld->wireaddrs[n].addrlen = TOR_V2_ADDRLEN;
+	reach->ld->wireaddrs[n].port = reach->ld->portnum;
+	b32_decode((u8 *)tor_dec_bytes,(u8 *)reach->buffer,2);
```

Comment:
Would it not be possible to use `parse_wireaddr` here?  That way we can keep `b32_encode` and `b32_decode` private to `common/tor.c`.